persistence.js

```diff
@@ -9,348 +7,318 @@ const QlobberOpts = {
-  const toDelete = store[packet.messageId]
-  let err = null
+    if (!toDelete) {
+      err = new Error('no such packet')
+    } else {
+      delete store[packet.messageId]
+    }
 
-  if (!toDelete) {
-    err = new Error('no such packet')
-  } else {
-    delete store[packet.messageId]
+    cb(err)
   }
 
-  cb(err)
-}
+  putWill (client, packet, cb) {
+    packet.brokerId = this.broker.id
+    packet.clientId = client.id
+    this._wills.set(client.id, packet)
+    cb(null, client)
+  }
 
-MemoryPersistence.prototype.putWill = function (client, packet, cb) {
-  packet.brokerId = this.broker.id
-  packet.clientId = client.id
-  this._wills[client.id] = packet
-  cb(null, client)
-}
+  getWill (client, cb) {
+    cb(null, this._wills.get(client.id), client)
+  }
 
-MemoryPersistence.prototype.getWill = function (client, cb) {
-  cb(null, this._wills[client.id], client)
-}
+  delWill (client, cb) {
+    const will = this._wills.get(client.id)
+    this._wills.delete(client.id)
+    cb(null, will, client)
+  }
 
-MemoryPersistence.prototype.delWill = function (client, cb) {
-  const will = this._wills[client.id]
-  delete this._wills[client.id]
-  cb(null, will, client)
-}
+  streamWill (brokers = {}) {
+    return Readable.from(willsByBrokers(this._wills, brokers))
+  }
 
-MemoryPersistence.prototype.streamWill = function (brokers) {
-  const clients = Object.keys(this._wills)
-  const wills = this._wills
-  brokers = brokers || {}
-  return from2.obj(function match (size, next) {
-    let entry
-
-    while ((entry = clients.shift()) != null) {
-      if (!brokers[wills[entry].brokerId]) {
-        setImmediate(next, null, wills[entry])
-        return
-      }
-    }
+  getClientList (topic) {
+    return Readable.from(clientListbyTopic(this._subscriptions, topic))
+  }
 
-    if (!entry) {
-      this.push(null)
+  destroy (cb) {
+    this._retained = null
+    if (cb) {
+      cb(null)
     }
-  })
+  }
 }
 
-MemoryPersistence.prototype.getClientList = function (topic) {
-  const clientSubs = this._subscriptions
-  const entries = clientSubs.entries(clientSubs)
-  return from2.obj(function match (size, next) {
-    let entry
-    while (!(entry = entries.next()).done) {
-      if (entry.value[1].has(topic)) {
-        setImmediate(next, null, entry.value[0])
-        return
-      }
-    }
-    next(null, null)
-  })
+function _outgoingEnqueue (sub, packet) {
+  const id = sub.clientId
+  const queue = getMapRef(this._outgoing, id, [], CREATE_ON_EMPTY)
+  queue[queue.length] = new Packet(packet)
 }
 
-MemoryPersistence.prototype.destroy = function (cb) {
-  this._retained = null
-  if (cb) {
-    cb(null)
+function getMapRef (map, key, ifEmpty, createOnEmpty = false) {
+  let value = map.get(key)
+  if (value === undefined && createOnEmpty) {
+    value = ifEmpty
+    map.set(key, value)
   }
+  return value
 }
 
-module.exports = MemoryPersistence
+module.exports = () => { return new MemoryPersistence() }
```

Comment:
Could you explain me why this?

Reply:
Of course :-)

On pre-ES6 classes the constructor also doubles as a regular function so a new instance can be created by just calling the function eg.:
```javascript
const instance = MemoryPersistence()
```
This is also why the current version starts with: https://github.com/moscajs/aedes-persistence/blob/883c249cac14168332e078f38037c605bcc18089/persistence.js#L13-L17
On ES6 classes you can't do this:
```javascript
const instance = MemoryPersistence()
```
As the class name refers to the class and not the constructor.
So we need to do:

```javascript
const instance = new MemoryPersistence()
```
Which will invoke the constructor.

Now at this point all our users will expect us to provide a direct callable export.
Hence we need to export:
```javascript
module.exports = () => { return new MemoryPersistence() }
```
So any importer can still do:
```javascript
const instance = MemoryPersistence()
```
Ideally we would also export the class itself by adding:
```javascript
module.exports.MemoryPersistence = MemoryPersistence
```
so future importers can do:
```javascript
const { MemoryPersistence }  = require ("aedes-persistence")
const persistence = new MemoryPersistence()
```
Which would make it more clear that `persistence` is a class instance.
I added this last export at first but it lowered the coveralls score as we don't have any code using this new export. So I pulled it out ;-)

Hope this clarifies it a bit.
Feel free to ask more questions !

Kind regards,
Hans